File: backend/assistant/store.py
"""
Conversation storage for the dashboard assistant.

Threads and messages, in the same MongoDB database as everything else. Two
things here are worth knowing.

**Every read is scoped by customer_id.** Not "the caller should pass the right
id" — the queries take it and filter on it, so there is no accessor that can
return someone else's conversation by being called carelessly.

**History is capped when it is read, not when it is written.** The full thread
is kept for the person to scroll; only the last N turns are handed to the
model. Those are different jobs and a single limit would do one of them badly.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pymongo import ASCENDING, DESCENDING, ReturnDocument

from backend.shared import config
from backend.shared.mongo import (
    coll,
    document,
    documents,
    object_id,
    register_indexes,
    to_object_id,
)

ROLE_USER = "user"
ROLE_ASSISTANT = "assistant"

THREADS = "assistant_threads"
MESSAGES = "assistant_messages"

register_indexes(THREADS, [
    ([("customer_id", ASCENDING), ("updated_at", DESCENDING)], {"name": "by_customer"}),
])
register_indexes(MESSAGES, [
    ([("thread_id", ASCENDING), ("_id", ASCENDING)], {"name": "by_thread"}),
    # count_messages_today filters on all three at once.
    ([("customer_id", ASCENDING), ("role", ASCENDING), ("created_at", ASCENDING)],
     {"name": "by_customer_role_day"}),
])


def _now() -> str:
    return datetime.now(config.IST).isoformat()


def init_assistant_tables() -> None:
    """Kept as an entry point for startup; MongoDB needs no schema built."""
    return None


# ---------------------------------------------------------------------------
# Threads
# ---------------------------------------------------------------------------

def create_thread(customer_id: Any, title: str = "New chat") -> dict[str, Any]:
    now = _now()
    doc = {
        "customer_id": to_object_id(customer_id),
        "title": title[:120] or "New chat",
        "created_at": now,
        "updated_at": now,
    }
    result = coll(THREADS).insert_one(doc)
    return document({**doc, "_id": result.inserted_id})  # type: ignore[return-value]


def get_thread(customer_id: Any, thread_id: Any) -> dict[str, Any] | None:
    """One thread, or None. Scoped — another customer's id simply doesn't match."""
    thread_oid = object_id(thread_id)
    customer_oid = object_id(customer_id)
    if thread_oid is None or customer_oid is None:
        return None
    return document(
        coll(THREADS).find_one({"_id": thread_oid, "customer_id": customer_oid})
    )


def list_threads(customer_id: Any, limit: int = 50) -> list[dict[str, Any]]:
    oid = object_id(customer_id)
    if oid is None:
        return []
    return documents(
        coll(THREADS)
        .find({"customer_id": oid})
        .sort("updated_at", DESCENDING)
        .limit(limit)
    )


def rename_thread(customer_id: Any, thread_id: Any, title: str) -> dict[str, Any] | None:
    thread_oid = object_id(thread_id)
    customer_oid = object_id(customer_id)
    if thread_oid is None or customer_oid is None:
        return None

    # Scoped in the update itself rather than checked first, so there is no
    # window in which the ownership test and the write disagree.
    return document(coll(THREADS).find_one_and_update(
        {"_id": thread_oid, "customer_id": customer_oid},
        {"$set": {"title": title[:120] or "New chat", "updated_at": _now()}},
        return_document=ReturnDocument.AFTER,
    ))


def delete_thread(customer_id: Any, thread_id: Any) -> bool:
    thread_oid = object_id(thread_id)
    customer_oid = object_id(customer_id)
    if thread_oid is None or customer_oid is None:
        return False

    deleted = coll(THREADS).delete_one(
        {"_id": thread_oid, "customer_id": customer_oid}
    ).deleted_count

    # Messages are removed explicitly. MongoDB has no cascade, and a thread
    # whose messages outlived it would be invisible but still stored.
    if deleted:
        coll(MESSAGES).delete_many({"thread_id": thread_oid})
    return bool(deleted)


# ---------------------------------------------------------------------------
# Messages
# ---------------------------------------------------------------------------

def add_message(
    customer_id: Any, thread_id: Any, role: str, content: str
) -> dict[str, Any] | None:
    """
    Append a message, if the thread belongs to *customer_id*.

    Returns None when it doesn't, rather than writing into someone else's
    conversation.
    """
    if get_thread(customer_id, thread_id) is None:
        return None

    now = _now()
    doc = {
        "thread_id": to_object_id(thread_id),
        # Denormalised so "how many did this person send today?" is one indexed
        # read instead of a join. Messages are never reassigned to another
        # customer, so the copy cannot drift from the thread it belongs to.
        "customer_id": to_object_id(customer_id),
        "role": role,
        "content": content,
        "created_at": now,
    }
    result = coll(MESSAGES).insert_one(doc)
    coll(THREADS).update_one(
        {"_id": to_object_id(thread_id)}, {"$set": {"updated_at": now}}
    )
    return document({**doc, "_id": result.inserted_id})


def list_messages(customer_id: Any, thread_id: Any) -> list[dict[str, Any]]:
    """The whole conversation, for display."""
    if get_thread(customer_id, thread_id) is None:
        return []
    return documents(
        coll(MESSAGES).find({"thread_id": object_id(thread_id)}).sort("_id", ASCENDING)
    )


def history_for_model(customer_id: Any, thread_id: Any) -> list[dict[str, str]]:
    """
    The tail of the conversation, shaped for Ollama's ``messages`` array.

    Capped at ``ASSISTANT_HISTORY_TURNS`` because every extra turn is more
    prompt for a CPU to re-read on each reply, and a 7B model loses the thread
    long before it runs out of context window.
    """
    messages = list_messages(customer_id, thread_id)
    tail = messages[-config.ASSISTANT_HISTORY_TURNS:]
    return [{"role": message["role"], "content": message["content"]} for message in tail]


# ---------------------------------------------------------------------------
# Usage
# ---------------------------------------------------------------------------

def count_messages_today(customer_id: Any) -> int:
    """
    How many questions this customer has asked today (IST).

    Generation on two ARM cores is the scarcest thing this box has, so it is
    rationed per person rather than left open.
    """
    oid = object_id(customer_id)
    if oid is None:
        return 0

    today = datetime.now(config.IST).date().isoformat()
    return coll(MESSAGES).count_documents({
        "customer_id": oid,
        "role": ROLE_USER,
        # created_at is an ISO timestamp, so "today" is a prefix range. Stated
        # as a range rather than a regex so the index can serve it.
        "created_at": {"$gte": today, "$lt": today + "￿"},
    })

File: node-backend/src/assistant/store.ts
/**
 * Conversation storage for the dashboard assistant.
 *
 * Port of `backend/assistant/store.py`.
 *
 * Threads and messages, in the same MongoDB database as everything else. Two
 * things here are worth knowing.
 *
 * **Every read is scoped by customer_id.** Not "the caller should pass the
 * right id" — the queries take it and filter on it, so there is no accessor
 * that can return someone else's conversation by being called carelessly.
 *
 * **History is capped when it is read, not when it is written.** The full
 * thread is kept for the person to scroll; only the last N turns are handed to
 * the model. Those are different jobs and a single limit would do one of them
 * badly.
 */

import { ASSISTANT_HISTORY_TURNS } from "../config.js";
import {
  coll,
  document,
  documents,
  objectId,
  registerIndexes,
  toObjectId,
  type Doc,
} from "../shared/mongo.js";
import { nowISO, todayIST } from "../shared/time.js";

export const ROLE_USER = "user";
export const ROLE_ASSISTANT = "assistant";

export const THREADS = "assistant_threads";
export const MESSAGES = "assistant_messages";

registerIndexes(THREADS, [
  [{ customer_id: 1, updated_at: -1 }, { name: "by_customer" }],
]);

registerIndexes(MESSAGES, [
  [{ thread_id: 1, _id: 1 }, { name: "by_thread" }],
  // countMessagesToday filters on all three at once.
  [
    { customer_id: 1, role: 1, created_at: 1 },
    { name: "by_customer_role_day" },
  ],
]);

/** Kept as an entry point for startup; MongoDB needs no schema built. */
export function initAssistantTables(): void {
  return;
}

// ---------------------------------------------------------------------------
// Threads
// ---------------------------------------------------------------------------

export async function createThread(customerId: unknown, title = "New chat"): Promise<Doc> {
  const now = nowISO();
  const doc = {
    customer_id: toObjectId(customerId),
    title: title.slice(0, 120) || "New chat",
    created_at: now,
    updated_at: now,
  };

  const threads = await coll(THREADS);
  const result = await threads.insertOne({ ...doc });
  return document({ ...doc, _id: result.insertedId })!;
}

/** One thread, or null. Scoped — another customer's id simply doesn't match. */
export async function getThread(customerId: unknown, threadId: unknown): Promise<Doc | null> {
  const threadOid = objectId(threadId);
  const customerOid = objectId(customerId);
  if (threadOid === null || customerOid === null) return null;

  const threads = await coll(THREADS);
  return document(await threads.findOne({ _id: threadOid, customer_id: customerOid }));
}

export async function listThreads(customerId: unknown, limit = 50): Promise<Doc[]> {
  const oid = objectId(customerId);
  if (oid === null) return [];

  const threads = await coll(THREADS);
  return documents(
    await threads.find({ customer_id: oid }).sort({ updated_at: -1 }).limit(limit).toArray(),
  );
}

export async function renameThread(
  customerId: unknown,
  threadId: unknown,
  title: string,
): Promise<Doc | null> {
  const threadOid = objectId(threadId);
  const customerOid = objectId(customerId);
  if (threadOid === null || customerOid === null) return null;

  // Scoped in the update itself rather than checked first, so there is no
  // window in which the ownership test and the write disagree.
  const threads = await coll(THREADS);
  return document(
    await threads.findOneAndUpdate(
      { _id: threadOid, customer_id: customerOid },
      { $set: { title: title.slice(0, 120) || "New chat", updated_at: nowISO() } },
      { returnDocument: "after" },
    ),
  );
}

export async function deleteThread(customerId: unknown, threadId: unknown): Promise<boolean> {
  const threadOid = objectId(threadId);
  const customerOid = objectId(customerId);
  if (threadOid === null || customerOid === null) return false;

  const threads = await coll(THREADS);
  const deleted = (await threads.deleteOne({ _id: threadOid, customer_id: customerOid }))
    .deletedCount;

  // Messages are removed explicitly. MongoDB has no cascade, and a thread whose
  // messages outlived it would be invisible but still stored.
  if (deleted) {
    const messages = await coll(MESSAGES);
    await messages.deleteMany({ thread_id: threadOid });
  }
  return Boolean(deleted);
}

// ---------------------------------------------------------------------------
// Messages
// ---------------------------------------------------------------------------

/**
 * Append a message, if the thread belongs to *customerId*.
 *
 * Returns null when it doesn't, rather than writing into someone else's
 * conversation.
 */
export async function addMessage(
  customerId: unknown,
  threadId: unknown,
  role: string,
  content: string,
): Promise<Doc | null> {
  if ((await getThread(customerId, threadId)) === null) return null;

  const now = nowISO();
  const doc = {
    thread_id: toObjectId(threadId),
    // Denormalised so "how many did this person send today?" is one indexed
    // read instead of a join. Messages are never reassigned to another
    // customer, so the copy cannot drift from the thread it belongs to.
    customer_id: toObjectId(customerId),
    role,
    content,
    created_at: now,
  };

  const messages = await coll(MESSAGES);
  const result = await messages.insertOne({ ...doc });

  const threads = await coll(THREADS);
  await threads.updateOne({ _id: toObjectId(threadId) }, { $set: { updated_at: now } });

  return document({ ...doc, _id: result.insertedId });
}

/** The whole conversation, for display. */
export async function listMessages(customerId: unknown, threadId: unknown): Promise<Doc[]> {
  if ((await getThread(customerId, threadId)) === null) return [];

  const messages = await coll(MESSAGES);
  return documents(
    await messages.find({ thread_id: objectId(threadId) }).sort({ _id: 1 }).toArray(),
  );
}

/**
 * The tail of the conversation, shaped for the model's `messages` array.
 *
 * Capped at `ASSISTANT_HISTORY_TURNS` because every extra turn is more prompt
 * for a CPU to re-read on each reply, and a 7B model loses the thread long
 * before it runs out of context window.
 */
export async function historyForModel(
  customerId: unknown,
  threadId: unknown,
): Promise<Array<{ role: string; content: string }>> {
  const messages = await listMessages(customerId, threadId);
  const tail = messages.slice(-ASSISTANT_HISTORY_TURNS);
  return tail.map((message) => ({ role: message.role, content: message.content }));
}

// ---------------------------------------------------------------------------
// Usage
// ---------------------------------------------------------------------------

/**
 * How many questions this customer has asked today (IST).
 *
 * Generation on two ARM cores is the scarcest thing this box has, so it is
 * rationed per person rather than left open.
 */
export async function countMessagesToday(customerId: unknown): Promise<number> {
  const oid = objectId(customerId);
  if (oid === null) return 0;

  const today = todayIST();
  const messages = await coll(MESSAGES);
  return messages.countDocuments({
    customer_id: oid,
    role: ROLE_USER,
    // created_at is an ISO timestamp, so "today" is a prefix range. Stated as a
    // range rather than a regex so the index can serve it. U+FFFF sorts above
    // every character a timestamp can contain.
    created_at: { $gte: today, $lt: today + "￿" },
  });
}
